package com.priynshuchouhn.learn_spring_framework.examples.a1;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.AnnotationConfigApplicationContext;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;
import org.springframework.stereotype.Component;

import java.util.Arrays;

@Component
class YourBusiness{
    Dependency1 dependency1;
    Dependency2 dependency2;

    public YourBusiness(Dependency1 dependency1, Dependency2 dependency2) {
        System.out.println("constructor injection - dependency1: " + dependency1 +  ", dependency2: " + dependency2);
        this.dependency1 = dependency1;
        this.dependency2 = dependency2;
    }
//    @Autowired
//    public void setDependency1(Dependency1 dependency1) {
//        System.out.println("setter injection - dependency1");
//        this.dependency1 = dependency1;
//    }

//    @Autowired
//    public void setDependency2(Dependency2 dependency2) {
//        System.out.println("setter injection - dependency2");
//        this.dependency2 = dependency2;
//    }

    public String toString() {
        return "using "+ dependency1 + " and "+dependency2;
    }
}

@Component
class Dependency1 {}
@Component
class Dependency2{}

@Configuration
@ComponentScan
public class DepInjectionLauncherApplication {
    public static void main(String[] args) {
        var context = new AnnotationConfigApplicationContext(DepInjectionLauncherApplication.class);
        Arrays.stream(context.getBeanDefinitionNames()).forEach(System.out::println);
        System.out.println(context.getBean(YourBusiness.class));
    }

}
